Sources/SemanticIndex/UpToDateTracker.swift

```diff
@@ -12,58 +12,77 @@
 
 import Foundation
 
-/// Keeps track of whether an item (a target or file to index) is up-to-date.
-actor UpToDateTracker<Item: Hashable> {
-  private enum Status {
-    /// The item is up-to-date.
-    case upToDate
+struct DummySecondaryKey: Hashable {}
 
-    /// The target or file has been marked out-of-date at the given date.
+/// Keeps track of whether an item (a target or file to index) is up-to-date.
+///
+/// The `UpToDateTracker` has two sets of keys. The primary key is the key by which items in the `UpToDateTracker` can
+/// be marked as out of date, but only a combination of primary key and secondary key can be marked as up-to-date.
+/// This is useful so that we invalidate the index status for a source file's URI (the primary key) when the file
+/// receives an update but only mark it up-to-date in with respect to a target (secondary key).
+///
+/// `DummySecondaryKey` can be used as `SecondaryKey` if the secondary key tracking is not needed.
+actor UpToDateTracker<PrimaryKey: Hashable, SecondaryKey: Hashable> {
+  private struct Status {
+    /// The date at which this primary key has last been marked out-of-date or `nil` if it has never been marked
+    /// out-of-date.
     ///
     /// Keeping track of the date is necessary so that we don't mark a target as up-to-date if we have the following
     /// ordering of events:
     ///  - Preparation started
     ///  - Target marked out of date
     ///  - Preparation finished
-    case outOfDate(Date)
+    var lastOutOfDate: Date?
+
+    /// The secondary keys for which the item is considered up-to-date.
+    var secondaryKeys: Set<SecondaryKey>
```

Comment:
This probably has to end up becoming more general. Consider more accurate tracking for preparation - there we need to know the files that were changed and their language, since the target is up to date if you're in a swift file and only swift files have changed but not up to date if you're in a objc file and swift has changed. In that case the check isn't a "do we have an up to date secondary key", but instead a check over the files that have changed (or some status tracking that in a more general manner, since we don't actually care about the exact files).

Happy to leave this as is for now and consider that case when we get to it though.

Reply:
Yes, let’s re-consider this when we implement file-level preparation tracking and get and fully understand what we need here.